src/components/application_manager/src/app_service_manager.cc

```diff
@@ -528,4 +532,256 @@ void AppServiceManager::AppServiceUpdated(
+              std::make_shared<smart_objects::SmartObject>(
+                  current_request_[correlation_id].message);
+          AddTimeoutTimer(correlation_id);
+          app_manager_.GetRPCService().SendMessageToMobile(result);
+        }
+        return true;
+      }
+
+      LOG4CXX_DEBUG(logger_, "Valid RPC passing response");
+
+      if (current_request_.find(correlation_id) != current_request_.end()) {
+        LOG4CXX_DEBUG(
+            logger_,
+            "Forwarding response to mobile app "
+                << current_request_[correlation_id].origin_connection_key);
+        rpc_message[strings::params][strings::connection_key] =
+            current_request_[correlation_id].origin_connection_key;
+        smart_objects::SmartObjectSPtr result =
+            std::make_shared<smart_objects::SmartObject>(rpc_message);
+
+        request_queue_.erase(correlation_id);
+        current_request_.erase(correlation_id);
+        RemoveTimeoutTimer(correlation_id);
+        app_manager_.GetRPCService().SendMessageToMobile(result);
+        return true;
+      }
+
+      return false;
+    } break;
+  }
+
+  return false;
+}
+
+bool RPCPassingHandler::CycleThroughRPCPassingRequests(
+    uint32_t correlation_id) {
+  LOG4CXX_DEBUG(logger_, "Cycling through pass through requests");
+  if (request_queue_.find(correlation_id) == request_queue_.end()) {
+    LOG4CXX_ERROR(logger_, "Correlation id does NOT exist in map");
+    return false;
+  }
+
+  current_request_[correlation_id] = request_queue_[correlation_id].front();
+  request_queue_[correlation_id].pop();
+
+  MessageHelper::PrintSmartObject(current_request_[correlation_id].message);
+  return true;
+}
+
+void RPCPassingHandler::OnPassThroughRequestTimeout() {
+  LOG4CXX_DEBUG(logger_, "RPC Passing request timeout");
+  auto timeout_entry = timeout_queue_.front();
+  uint32_t correlation_id = std::get<1>(timeout_entry);
+  timeout_queue_.pop();
+  LOG4CXX_DEBUG(logger_,
+                "Cycling through requests for correlation_id "
+                    << correlation_id);
+  CycleThroughRPCPassingRequests(correlation_id);
+}
+
+void RPCPassingHandler::AddTimeoutTimer(uint32_t correlation_id) {
+  TimerSPtr rpc_passing_timer(std::make_shared<timer::Timer>(
+      "RPCPassingTimeoutTimer",
+      new timer::TimerTaskImpl<RPCPassingHandler>(
+          this, &RPCPassingHandler::OnPassThroughRequestTimeout)));
+  const uint32_t timeout_ms =
+      app_manager_.get_settings().rpc_pass_through_timeout();
+  rpc_passing_timer->Start(timeout_ms, timer::kSingleShot);
+  timeout_queue_.push(std::make_pair(rpc_passing_timer, correlation_id));
+}
+
+void RPCPassingHandler::RemoveTimeoutTimer(uint32_t correlation_id) {
+  auto timeout_entry = timeout_queue_.front();
+  TimerSPtr timer = std::get<0>(timeout_entry);
+  uint32_t cid = std::get<1>(timeout_entry);
+  if (cid != correlation_id) {
+    LOG4CXX_ERROR(logger_, "Deleting incorrect timer");
```

Comment:
You can't just remove the wrong timer, you need to actually search for the correct timer to remove.

Reply:
Fixed 👍